pkg/controller/node/ipam/cloud_cidr_allocator.go

```diff
@@ -105,6 +106,12 @@ func NewCloudCIDRAllocator(client clientset.Interface, cloud cloudprovider.Inter
 			if newNode.Spec.PodCIDR == "" {
 				return ca.AllocateOrOccupyCIDR(newNode)
 			}
+			// Even if PodCIDR is assigned, but NetworkUnavailable condition is
+			// set to true, we need to process the node to set the condition.
+			_, cond := v1node.GetNodeCondition(&newNode.Status, v1.NodeNetworkUnavailable)
+			if cond == nil || cond.Status != v1.ConditionFalse {
+				return ca.AllocateOrOccupyCIDR(newNode)
```

Comment:
IIUC we can still let the condition remain unset due to race. Consider the following scenario:

- on the first call to `AllocateOrOccupyCIDR`, we set the CIDR and the node condition (but the function hasn't returned yet, so we haven't yet done `removeNodeFromProcessing` for the node)
- some change from somewhere else overwrites the node condition (the above function still didn't return)
- we get the watch event for the 2nd step and reach here.. enter `AllocateOrOccupyCIDR`  (because the condition is lost), but not do anything as the node is still in the processing

We probably can live with this? (assuming that there will be some node update soon'ish which would fix it)

Reply:
Yup - this is completely fine, given that node updates are pretty frequent by design. This can introduce 10s delay or so, but that's not a big deal - it will be fixed eventually.

Reply:
That sounds reasonable. Though I think it is pretty easy to fix this :)
Just move the node condition setting part to a separate function and call that directly from here.

Reply:
We don't want to have API calls in event handler. And I would like to avoid larger changes during code freeze.

Reply:
SG